Extend read_vtk to add FIELD data

OpenFCST opened the pull request:
This PR extends the functionality of read_vtk to include FIELD data which can be used to add further information to each cell beyond material ID. In OpenFCST this is used for example to store pore size in each cell. To access the field data, please use the function get_field_data which returns a map with a key (in the form of a string) and then a map with cellID and value.

**NOTE**: This change was previously submitted as a different PR (#8575), but we decided to create a new PR because 1) we were asked to rebase to the current master, and this resulted in some conflicts which needed manual merging and 2) our commits in the previous PR did not conform to the author naming convention used in deal.ii, which seems to be a requirement in order to run the 'make indent' utility (This was also one of the review comments, since our indentation style was different from deal.ii). All the review comments from the old PR have now been implemented here.

masterleinad commented:
> 2) our commits in the previous PR did not conform to the author naming convention used in deal.ii, which seems to be a requirement in order to run the 'make indent' utility (This was also one of the review comments, since our indentation style was different from deal.ii).

:slightly_smiling_face: Well, CI is still showing
```
invalid author 'OpenFCST' without firstname and lastname

hint: for possible solutions, consult the webpage:
      https://github.com/dealii/dealii/wiki/Indentation#commit-authorship
##[error]Process completed with exit code 2.
```

masterleinad commented on include/deal.II/grid/grid_in.h:
```diff
@@ -670,6 +671,20 @@ class GridIn
+   * To access the value use field_data[name_field][cell_id].
+   */
+  const std::map<std::string, std::vector<double>> &
+  get_field_data()
```
```suggestion
  get_field_data() const
```
Also, we define member functions outside the class if possible.

bangerth commented:
@OpenFCST Any news?

OpenFCST commented:
> @OpenFCST Any news?

We have addressed all comments, however we do not know how to solve the issue with authorship that make the indent test fail. Any ideas on how to address those?

bangerth commented:
Take a look here: https://github.com/dealii/dealii/wiki/Indentation#commit-authorship

masterleinad commented:
/rebuild

peterrum commented:
Let's tackle this after the release. I am adjusting the milestone here.

vaishnavi-kale commented:
Dear deal.ii developers, 

Could you please help us resolve the CI issues on this PR? Our CI tests are failing because of an invalid author name 'OpenFCST'. We followed the procedure in the guide (https://github.com/dealii/dealii/wiki/Indentation#commit-authorship) to ensure that our contributors' author names followed the required convention. However, the repository owner is 'OpenFCST', and creating a fork / submitting a PR from our fork to the deal.ii github repository can only be done by this user. This is creating commits with the invalid author name (For example, this typically happens while syncing our master with the deal.ii master from the github webpage, which creates a new commit). We tried changing the username for OpenFCST by editing the username in the 'profile' tab, but the issue was not resolved by doing this.  

PS: We also tried rewriting commit history using the git filter-branch method, but this created over 20,000 new commits, and CI tests did not run on this branch.

drwells commented:
You can always set authorship information from the command line like

> git commit -m "Message" --author="John Smith <john@smith.com>"

which will satisfy our authorship tracking requirements. We check author names, not commiters, so this should be OK even if the organization or account name is different. It sounds like you aren't using the command line - I don't know if there is a way to set authorship information from one of the wrapper apps.

Independent of that - we need to do some more work in the rebase as whatever changes were made to extend read_vtk are now mixed in with a lot of irrelevant changes. Do you still have the original branch (alternatively, is it present in reflog)? I believe the best way forward would be to rebase those changes on top of deal.II's master branch and not use the present commit history.

vaishnavi-kale commented:
Yes, we are facing this issue only when performing actions in the Github web browser.  For instance, when we work on a feature, this is done locally in a separate branch where we use terminal commands to commit, set author names etc. But after the local changes in the feature branch are pushed to Github, the repo owner (OpenFCST) must accept a PR to merge the feature branch to the fork/master or to submit a PR to deal.ii. This creates a new commit with an invalid author name.  

Regarding rebase to the new master - if there is no workaround to the Github author issue, we will add the changes manually to the latest deal.ii code and try to create a new PR from the terminal instead of the web browser.

drwells commented:
There are a couple of possible workarounds:
1. I checked out your branch: only the first commit is from OpenFCST and that's an artificial commit which should not exist. The second commit is yours and has correct authorship information. Since we ignore merge commits when checking authorship, the process of merging into an OpenFCST branch and then opening a PR from that branch here should be OK.
2. Could you submit this PR from your own fork of deal.II instead of OpenFCST/dealii instead? That would make this management a lot easier because there is simply no need to go through the other account - we are just changing dealii after all, not OpenFCST.

drwells commented:
Trying to create a PR from OpenFCST/dealii cannot work since our commit histories are out of sync: see, e.g.,

https://github.com/OpenFCST/dealii/commit/0a6f3c3ad63361d86e40e671d6686bed96487809

Hence I think cherry-picking commit 442ed0fad421350a890fc324b511e9260d44ce93 onto a new up-to-date branch (branched off of dealii/dealii's master branch) is the best way forward.

bangerth commented:
@OpenFCST What should we do with this PR?

drwells commented:
It's been 9 months without an update - lets close this for now. We can always reopen this PR if there's some forward progress from the FCST team.
